### PR TITLE
SONARJS-899 Improve InfiniteLoopCheck: account for simple value ranges

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-checks/src/main/java/org/sonar/javascript/checks/LoopsShouldNotBeInfiniteCheck.java [hunk 1]
```diff
@@ -31,45 +34,101 @@
...
     }
+
+    return false;
+  }
+
+  private static IterationStatementTree getLoop(JavaScriptTree condition) {
+    return (IterationStatementTree) CheckUtils.getFirstAncestor(condition, Kind.FOR_STATEMENT, Kind.WHILE_STATEMENT, Kind.DO_WHILE_STATEMENT);
   }
 
   private static boolean isInfiniteLoop(IterationStatementTree tree, @Nullable JavaScriptTree condition) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/LoopsShouldNotBeInfiniteCheck.java]: MINOR Move this method into "TreeVisitor". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3398)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: javascript-checks/src/main/java/org/sonar/javascript/checks/LoopsShouldNotBeInfiniteCheck.java [hunk 2]
```diff
@@ -31,45 +34,101 @@
...
 import org.sonar.plugins.javascript.api.visitors.DoubleDispatchVisitor;
-import org.sonar.plugins.javascript.api.visitors.DoubleDispatchVisitorCheck;
 import org.sonar.plugins.javascript.api.visitors.IssueLocation;
 
 @Rule(key = "S2189")
-public class LoopsShouldNotBeInfiniteCheck extends DoubleDispatchVisitorCheck {
+public class LoopsShouldNotBeInfiniteCheck extends SeCheck {
+
+  private Set<Tree> loopConditions = new HashSet<>();
+  private TreeVisitor treeVisitor = new TreeVisitor();
```
Comment (inverno) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/LoopsShouldNotBeInfiniteCheck.java]: Non need to store the TreeVisitor in a field. You can inline this.
Alternatively you could move loopConditions inside treeVisitor and let TreeVisitor do the filtering of the conditions. See my next comment.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/LoopsShouldNotBeInfiniteCheck.java [hunk 3]
```diff
@@ -31,45 +34,101 @@
...
+    conditionsWithIssue.clear();
+
+    scriptTree.accept(treeVisitor);
   }
 
   @Override
-  public void visitWhileStatement(WhileStatementTree tree) {
-    visitLoop(tree, tree.condition());
-    super.visitWhileStatement(tree);
+  public void checkConditions(Map<Tree, Collection<Constraint>> conditions) {
```
Comment (inverno) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/LoopsShouldNotBeInfiniteCheck.java]: You may make things a bit more explicit regarding the role of loopConditions and the conditions provided by this method by delegating to TreeVisitor the filtering of the conditions and then iterate on those.
Something like `treeVisitor.retainOnlyLoopConditions(conditions)`.

**Pull request merged**